internal/provider_framework/util.go

```diff
@@ -0,0 +1,90 @@
+package provider_framework
+
+import (
+	"crypto/sha1"
+	"encoding/hex"
+	"encoding/pem"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func decodePEM(d *schema.ResourceData, pemKey, pemType string) (*pem.Block, error) {
+	block, _ := pem.Decode([]byte(d.Get(pemKey).(string)))
+	if block == nil {
+		return nil, fmt.Errorf("no PEM block found in %s", pemKey)
+	}
+	if pemType != "" && block.Type != pemType {
+		return nil, fmt.Errorf("invalid PEM type in %s: %s", pemKey, block.Type)
+	}
+
+	return block, nil
+}
+
+// hashForState computes the hexadecimal representation of the SHA1 checksum of a string.
+// This is used by most resources/data-sources here to compute their Unique Identifier (ID).
+func hashForState(value string) string {
+	if value == "" {
+		return ""
+	}
+	hash := sha1.Sum([]byte(strings.TrimSpace(value)))
+	return hex.EncodeToString(hash[:])
+}
+
+// overridableTimeFunc normally returns time.Now(),
+// but it is overridden during testing to simulate an arbitrary value of "now".
+var overridableTimeFunc = func() time.Time {
+	return time.Now()
+}
+
+// attrPathToString takes all the tftypes.AttributePathStep in a tftypes.AttributePath and concatenates them,
+// using `.` as separator.
+//
+// This should be used only when trying to "print out" a tftypes.AttributePath in a log or an error message.
+func attrPathToString(path *tftypes.AttributePath) string {
```

Comment:
Wondering if we want this func in the framework. Think it would be broadly useful.

Reply:
This is going to be part of the new `attr.Path` work that @bflad has defined lately.